src/compiler/llvm-to-backend/host/HostKernelWrapperPass.cpp

```diff
@@ -172,6 +182,7 @@ llvm::Function *makeWrapperFunction(llvm::Function &F, std::int64_t DynamicLocal
     replaceUsesOfGVWith(*Wrapper, cbs::NumGroupsGlobalNames[I], NumGroups[I]);
     replaceUsesOfGVWith(*Wrapper, cbs::GroupIdGlobalNames[I], GroupIds[I]);
     replaceUsesOfGVWith(*Wrapper, cbs::LocalSizeGlobalNames[I], LocalSize[I]);
+    replaceUsesOfGVWith(*Wrapper, cbs::LocalIdGlobalNames[I], LocalIds[I]);
```

Comment:
I'm not sure if I understand what you are trying to achieve here / why this is necessary.
Is this an attempt to 0 initialize the `__acpp_cbs_local_id_[x-z]`?
Wouldn't this override all uses to `0` instead?

On the other hand, I'm not even sure, why there should any uses of these variables be around at this point.. 🤔 
Maybe the correct fix would be to fix the uses to be deleted correctly, beforehand.

Maybe this should be `mergeGVLoadsInEntry` instead:
https://github.com/AdaptiveCpp/AdaptiveCpp/blob/bf218d9b835efe2ac3c609a10abb1b997b00b116/src/compiler/cbs/SubCfgFormation.cpp#L1339
So that the, then unique loads, can be effectively deleted here: https://github.com/AdaptiveCpp/AdaptiveCpp/blob/bf218d9b835efe2ac3c609a10abb1b997b00b116/src/compiler/cbs/SubCfgFormation.cpp#L1355

And we should also delete those loads here as well:
https://github.com/AdaptiveCpp/AdaptiveCpp/blob/bf218d9b835efe2ac3c609a10abb1b997b00b116/src/compiler/cbs/SubCfgFormation.cpp#L1294

Or, potentially, I'm misunderstanding what is happening here/why this should be necessary :)

Reply:
I think the motivation is that the SSCP group algorithms interface is unaware of the dimensionality of the kernel. So often times we need to calculate the general linearized local id `lid_z * size_x * size*y + lid_y * size_x + lid_x`.

The problem is that the CBS builtins for the local id remain undefined if the dimension is greater than the dimensionality of the kernel, which creates linker errors.

So we need to have all builtin local ids defined always, independently from the dimensionality of the kernels. The ones that are outside the dimensionality of the kernel can just be set to constant 0. 

The idea behind @sbalint98's change here is that references to local ids that are still in the code after the CBS pipeline  has run must be cases where the dimension is outside the dimensionality of the kernel, and can thus be set to 0.
But it's not clear whether this is the best way to handle this :)

Reply:
`__acpp_cbs_local_id_[x-z]` is around because in the builtins need to know the local_id, however the dimensionality  is only available in runtime. Therefore we call  __acpp_sscp_typed_get_local_linear_id<3, int>()` which uses `__acpp_cbs_local_id_[x-z]`. In case the kernel is not 3D the unused `__acpp_cbs_local_id_[x-z]` references remain in the IR and cause a JIT time error.

Maybe we could just set these to zero in the SubCfgFormation createLoopsAroundKernel function? Or would deleting them have the same effect?

Reply:
Same happens with ` ACPP_ADAPTIVITY_LEVEL=0 ` unfortunately.

Reply:
Hm okay, I think, I understand the issue then.

Can you test if applying the following patch fixes this, without your compiler change?

```diff
diff --git a/src/compiler/cbs/SubCfgFormation.cpp b/src/compiler/cbs/SubCfgFormation.cpp
index 9b73472c..c62efdcf 100644
--- a/src/compiler/cbs/SubCfgFormation.cpp
+++ b/src/compiler/cbs/SubCfgFormation.cpp
@@ -351,6 +351,12 @@ void createLoopsAround(llvm::Function &F, llvm::BasicBlock *AfterBB,
 
   VMap[mergeGVLoadsInEntry(F, LocalIdGlobalNamesRotated[0])] = IndVars[0];
 
+  // in case code references all dimensions, we need to set the remaining dimensions to 0
+  for (size_t D = Dim; D < 3; ++D) {
+    VMap[mergeGVLoadsInEntry(F, LocalIdGlobalNamesRotated[D])] =
+        Builder.getIntN(Idx->getType()->getIntegerBitWidth(), 0);
+  }
+
   VMap[ContiguousIdx] = Idx;
   ContiguousIdx = Idx;
 }
@@ -1311,7 +1317,6 @@ void createLoopsAroundKernel(llvm::Function &F, llvm::DominatorTree &DT, llvm::L
 
   Body = Body->getSingleSuccessor();
 
-
   llvm::SmallVector<llvm::BasicBlock *, 4> ExitBBs;
   llvm::BasicBlock *ExitBB = llvm::BasicBlock::Create(F.getContext(), "exit", &F);
   llvm::IRBuilder<> Bld{ExitBB};
@@ -1352,10 +1357,13 @@ void createLoopsAroundKernel(llvm::Function &F, llvm::DominatorTree &DT, llvm::L
   llvm::remapInstructionsInBlocks(Blocks, VMap);
 
   // remove uses of the undefined global id variables
-  for (int D = 0; D < Dim; ++D)
-    if (auto *Load =
-            llvm::cast_or_null<llvm::LoadInst>(getLoadForGlobalVariable(F, LocalIdGlobalNames[D])))
+  for (int D = 0; D < 3; ++D)
+    if (auto *Load = llvm::cast_or_null<llvm::LoadInst>(
+            getLoadForGlobalVariable(F, LocalIdGlobalNames[D]))) {
+      if (D >= Dim)
+        Load->replaceAllUsesWith(llvm::ConstantInt::get(Load->getType(), 0));
       Load->eraseFromParent();
+    }
   HIPSYCL_DEBUG_EXECUTE_VERBOSE(F.viewCFG())
 }
 ```

Reply:
Hi, It seems to work as intended, do you want to push a commit somewhere that I could cherry pick for proper attribution?

Reply:
https://github.com/AdaptiveCpp/AdaptiveCpp/tree/cleanup-higher-dim-ids
Can you test this?
I'm not sure, I tested this a bit, and I'm not entirely sure, if it is sufficient..

Reply:
Interestingly this fails with:
```
[AdaptiveCpp Warning] /home/sbalint/.acpp/apps/sycl_tests-5273160718515756707/jit-cache/14570224624806148754.11565643320243800617.jit.so: undefined symbol: __acpp_cbs_local_id_y
```

Even though the symbols seem to be defined in the IR
```llvm
@__acpp_cbs_local_id_y = external local_unnamed_addr global i64, align 8
@__acpp_cbs_local_id_z = external local_unnamed_addr global i64, align 8

; Function Attrs: inaccessiblememonly mustprogress nocallback nofree nosync nounwind willreturn
declare void @llvm.assume(i1 noundef) #0

; Function Attrs: nofree nosync nounwind
define void @_Z18__acpp_sscp_kernelIN7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ25test_nd_group_function_1dILi102EiZN21group_functions_tests15group_broadcastIiE11test_methodEvEUlRSt6vectorIiSaIiEEmmE_ZNS7_11test_methodEvEUlT_mNS0_4sycl9sub_groupET0_iE_ZNS7_11test_methodEvEUlRKSA_SJ_mmmE_EvmT1_T2_T3_ENKUlRNSE_7handlerEE_clESP_EUlNSE_7nd_itemILi1EEEE_Li1EEEEvRKSD_(ptr nocapture readonly %0, ptr nocapture readonly %1) local_unnamed_addr #1 {
entry:
  %num_groups_x = load i64, ptr %0, align 8
  %2 = getelementptr inbounds { [3 x i64], [3 x i64], [3 x i64], ptr, ptr }, ptr %0, i64 0, i32 1, i64 0
  %group_id_x = load i64, ptr %2, align 8
  %3 = getelementptr inbounds { [3 x i64], [3 x i64], [3 x i64], ptr, ptr }, ptr %0, i64 0, i32 2, i64 0
  %local_size_x = load i64, ptr %3, align 8
  %4 = getelementptr inbounds { [3 x i64], [3 x i64], [3 x i64], ptr, ptr }, ptr %0, i64 0, i32 2, i64 1
  %local_size_y = load i64, ptr %4, align 8
  %5 = getelementptr inbounds { [3 x i64], [3 x i64], [3 x i64], ptr, ptr }, ptr %0, i64 0, i32 4
  %internal_local_mem_ptr = load ptr, ptr %5, align 8
  %6 = load ptr, ptr %1, align 8
  %7 = load ptr, ptr %6, align 8
  %8 = getelementptr inbounds ptr, ptr %1, i64 3
  %9 = load ptr, ptr %8, align 8
  %10 = load ptr, ptr %9, align 8
  %_alloca2.i = alloca i64, i64 %local_size_x, align 64, !hipSYCL.arrayified !372
  %11 = icmp ult i64 %num_groups_x, 8388608
  %12 = shl nuw nsw i64 %group_id_x, 8
  %13 = icmp ult i64 %group_id_x, 8388607
  %14 = load i64, ptr @__acpp_cbs_local_id_y, align 8
  %15 = trunc i64 %14 to i32
  %16 = load i64, ptr @__acpp_cbs_local_id_z, align 8
 %17 = trunc i64 %16 to i32
  %18 = trunc i64 %local_size_x to i32
  %19 = trunc i64 %local_size_y to i32
  %20 = mul i32 %17, %19
  %21 = add i32 %20, %15
  %22 = mul i32 %21, %18
  %umax = tail call i64 @llvm.umax.i64(i64 %local_size_x, i64 1)

```

Here is the full dump of the LLVM IR for all the stages:  [out.txt](https://github.com/user-attachments/files/18155053/out.txt)

Reply:
`external` means undefined :)
OK.. So doesn't really fix it.. I'll have a look tomorrow.

Reply:
Used the wrong name array :)
Should work now (did at least testing with your changes..)